core/connection.js

```diff
@@ -348,41 +410,29 @@ Blockly.Connection.prototype.isConnectionAllowed = function(candidate) {
 
   switch (candidate.type) {
     case Blockly.PREVIOUS_STATEMENT:
-      // Don't offer to connect the bottom of a statement block to one that's
-      // already connected.
-      if (candidate.isConnected() || this.isConnected()) {
-        return false;
-      }
-      break;
-    case Blockly.OUTPUT_VALUE:
+      return this.canConnectToPrevious_(candidate);
+    case Blockly.OUTPUT_VALUE: {
       // Don't offer to connect an already connected left (male) value plug to
       // an available right (female) value plug.
       if (candidate.isConnected() || this.isConnected()) {
         return false;
       }
       break;
-    case Blockly.INPUT_VALUE:
+    }
+    case Blockly.INPUT_VALUE: {
       // Offering to connect the left (male) of a value block to an already
       // connected value pair is ok, we'll splice it in.
-      // However, don't offer to splice into an unmovable block.
-      if (candidate.targetConnection &&
+      // However, don't offer to splice into an immovable block.
+      if (candidate.isConnected() &&
           !candidate.targetBlock().isMovable() &&
           !candidate.targetBlock().isShadow()) {
         return false;
       }
       break;
-    case Blockly.NEXT_STATEMENT:
-      // Don't let a block with no next connection bump other blocks out of the
-      // stack.  But covering up a shadow block or stack of shadow blocks is
-      // fine.  Similarly, replacing a terminal statement with another terminal
-      // statement is allowed.
-      if (candidate.isConnected() &&
-          !this.sourceBlock_.nextConnection &&
-          !candidate.targetBlock().isShadow() &&
-          candidate.targetBlock().nextConnection) {
-        return false;
-      }
+    }
+    case Blockly.NEXT_STATEMENT: {
```

Comment:
Did you remove all the checks for next_statement?

Reply:
Added back in a commit that I had forgotten to push.  It should be there now.